Adding check and test for robustBufferAccess usage

SzilardPapp-Arm opened the pull request:
Adding a warning if robustBufferAccess is enabled.
Warning message:
vkCreateDevice() called with enabled robustBufferAccess. Use robustBufferAccess as a debugging tool during development. Enabling it causes loss in performance for accesses to uniform buffers and shader storage buffers. Disable robustBufferAccess in release builds. Only leave it enabled if the application use-case requires the additional level of reliability due to the use of unverified user-supplied draw parameters.

mark-lunarg commented:
Running through CI.

mark-lunarg commented:
Hey @SzilardPapp-Arm, this test crashed on our Galaxy S10:

```
06-22 20:59:02.660 22587 22620 I VulkanLayerValidationTests: [ RUN      ] VkArmBestPracticesLayerTest.RobustBufferAccessTest
06-22 20:59:02.683 22587 22620 I vulkan  : Loaded layer VK_LAYER_KHRONOS_validation
06-22 20:59:02.758 22587 22620 W ValidationTest: 0xebadde09 skipped times: 0
--------- beginning of crash
06-22 20:59:02.758 22587 22620 F libc    : Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0 in tid 22620 (Thread-6), pid 22587 (ValidationTests)
06-22 20:59:02.798  5332  5744 I netd    : trafficSwapActiveStatsMap() <15.42ms>
06-22 20:59:02.813 29507 29507 E crash_dump64: unknown process state: t
06-22 20:59:02.839 29507 29507 I crash_dump64: obtaining output fd from tombstoned, type: kDebuggerdTombstone
06-22 20:59:02.840  5588  5588 I /system/bin/tombstoned: received crash request for pid 22620
06-22 20:59:02.843 29507 29507 I crash_dump64: performing dump of process 22587 (target tid = 22620)
06-22 20:59:02.852 29507 29507 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
06-22 20:59:02.853 29507 29507 F DEBUG   : Build fingerprint: 'samsung/beyond1ltexx/beyond1:10/QP1A.190711.020/G973FXXU3BSKO:user/release-keys'
06-22 20:59:02.853 29507 29507 F DEBUG   : Revision: '26'
06-22 20:59:02.853 29507 29507 F DEBUG   : ABI: 'arm64'
06-22 20:59:02.853 29507 29507 F DEBUG   : Timestamp: 2020-06-22 20:59:02-0600
06-22 20:59:02.853 29507 29507 F DEBUG   : pid: 22587, tid: 22620, name: Thread-6  >>> com.example.VulkanLayerValidationTests <<<
06-22 20:59:02.853 29507 29507 F DEBUG   : uid: 11913
06-22 20:59:02.853 29507 29507 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0
06-22 20:59:02.853 29507 29507 F DEBUG   : Cause: null pointer dereference
06-22 20:59:02.853 29507 29507 F DEBUG   :     x0  0000000000000001  x1  0000000000000000  x2  0000000000000000  x3  0000000000000000
06-22 20:59:02.853 29507 29507 F DEBUG   :     x4  0000000000000000  x5  0000000000000000  x6  0000000000000000  x7  000000702b2dbcd0
06-22 20:59:02.853 29507 29507 F DEBUG   :     x8  0000000000000001  x9  0000000000000000  x10 0000006fba5a1de8  x11 0000000000000028
06-22 20:59:02.853 29507 29507 F DEBUG   :     x12 0201010102010100  x13 000000702b2dc874  x14 000000702b2dc8b8  x15 000000003b9d71b0
06-22 20:59:02.853 29507 29507 F DEBUG   :     x16 0000000000000001  x17 0000000001020101  x18 000000003b9cf897  x19 0000006fb4336748
06-22 20:59:02.853 29507 29507 F DEBUG   :     x20 000000702b2dcc80  x21 0000000000000000  x22 0000000000000000  x23 0000000000000000
06-22 20:59:02.853 29507 29507 F DEBUG   :     x24 0000006fb4336740  x25 000000702b2de020  x26 0000006fb4336748  x27 000000702b2de020
06-22 20:59:02.853 29507 29507 F DEBUG   :     x28 000000702b2dc868  x29 000000702b2dd0c0
06-22 20:59:02.853 29507 29507 F DEBUG   :     sp  000000702b2dc850  lr  000000701c180870  pc  000000701c1809a8
06-22 20:59:02.863 29507 29507 F DEBUG   : 
06-22 20:59:02.863 29507 29507 F DEBUG   : backtrace:
06-22 20:59:02.864 29507 29507 F DEBUG   :       #00 pc 0000000000ab09a8  /vendor/lib64/egl/libGLES_mali.so (BuildId: 2074cca0c239c545a7bd2396c71003a9)
06-22 20:59:02.864 29507 29507 F DEBUG   :       #01 pc 0000000000ab05dc  /vendor/lib64/egl/libGLES_mali.so (vkCreateDevice+260) (BuildId: 2074cca0c239c545a7bd2396c71003a9)
06-22 20:59:02.864 29507 29507 F DEBUG   :       #02 pc 0000000000026f8c  /system/lib64/libvulkan.so (vulkan::driver::CreateDevice(VkPhysicalDevice_T*, VkDeviceCreateInfo const*, VkAllocationCallbacks const*, VkDevice_T**)+476) (BuildId: d569637329ba16330d390540dee0e601)
06-22 20:59:02.864 29507 29507 F DEBUG   :       #03 pc 00000000008dceb0  /data/app/com.example.VulkanLayerValidationTests-wGKoBqRYVR_wsQdtZ0wyDw==/lib/arm64/libVkLayer_khronos_validation.so (BuildId: 44934a49e4cdf81cc9e08014663b14d61afa4a41)
06-22 20:59:02.864 29507 29507 F DEBUG   :       #04 pc 00000000000164cc  /system/lib64/libvulkan.so (vulkan::api::CreateDevice(VkPhysicalDevice_T*, VkDeviceCreateInfo const*, VkAllocationCallbacks const*, VkDevice_T**)+1412) (BuildId: d569637329ba16330d390540dee0e601)
06-22 20:59:02.864 29507 29507 F DEBUG   :       #05 pc 00000000002fda38  /data/app/com.example.VulkanLayerValidationTests-wGKoBqRYVR_wsQdtZ0wyDw==/lib/arm64/libVulkanLayerValidationTests.so (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-22 20:59:02.864 29507 29507 F DEBUG   :       #06 pc 00000000003ab950  /data/app/com.example.VulkanLayerValidationTests-wGKoBqRYVR_wsQdtZ0wyDw==/lib/arm64/libVulkanLayerValidationTests.so (testing::Test::Run()+384) (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-22 20:59:02.864 29507 29507 F DEBUG   :       #07 pc 00000000003acc08  /data/app/com.example.VulkanLayerValidationTests-wGKoBqRYVR_wsQdtZ0wyDw==/lib/arm64/libVulkanLayerValidationTests.so (testing::TestInfo::Run()+396) (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-22 20:59:02.864 29507 29507 F DEBUG   :       #08 pc 00000000003ad1a8  /data/app/com.example.VulkanLayerValidationTests-wGKoBqRYVR_wsQdtZ0wyDw==/lib/arm64/libVulkanLayerValidationTests.so (testing::TestCase::Run()+248) (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-22 20:59:02.864 29507 29507 F DEBUG   :       #09 pc 00000000003b5f00  /data/app/com.example.VulkanLayerValidationTests-wGKoBqRYVR_wsQdtZ0wyDw==/lib/arm64/libVulkanLayerValidationTests.so (testing::internal::UnitTestImpl::RunAllTests()+1080) (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-22 20:59:02.864 29507 29507 F DEBUG   :       #10 pc 00000000003b5a94  /data/app/com.example.VulkanLayerValidationTests-wGKoBqRYVR_wsQdtZ0wyDw==/lib/arm64/libVulkanLayerValidationTests.so (testing::UnitTest::Run()+164) (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-22 20:59:02.864 29507 29507 F DEBUG   :       #11 pc 00000000001e09d4  /data/app/com.example.VulkanLayerValidationTests-wGKoBqRYVR_wsQdtZ0wyDw==/lib/arm64/libVulkanLayerValidationTests.so (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-22 20:59:02.864 29507 29507 F DEBUG   :       #12 pc 0000000000399aa8  /data/app/com.example.VulkanLayerValidationTests-wGKoBqRYVR_wsQdtZ0wyDw==/lib/arm64/libVulkanLayerValidationTests.so (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-22 20:59:02.864 29507 29507 F DEBUG   :       #13 pc 00000000000e3b14  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+36) (BuildId: 220051b49364b1c2da3adf10c30832cc)
06-22 20:59:02.864 29507 29507 F DEBUG   :       #14 pc 0000000000085330  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64) (BuildId: 220051b49364b1c2da3adf10c30832cc)
06-22 20:59:03.291  5588  5588 E /system/bin/tombstoned: Tombstone written to: /data/tombstones/tombstone_06
06-22 20:59:03.294  5811 29511 I DropBoxManagerService: add tag=data_app_native_crash isTagEnabled=true flags=0x2
```

SzilardPapp-Arm commented:
Hi @mark-lunarg,
I've added a check, hopefully that will resolve the problem.

mark-lunarg commented:
Running through CI again.

mark-lunarg commented:
Bummer, man!  Similar result...

```
06-24 09:27:59.201 15688 15728 I VulkanLayerValidationTests: [ RUN      ] VkArmBestPracticesLayerTest.RobustBufferAccessTest
06-24 09:27:59.223 15688 15728 I vulkan  : Loaded layer VK_LAYER_KHRONOS_validation
06-24 09:27:59.289 15688 15728 W ValidationTest: 0xebadde09 skipped times: 0
--------- beginning of crash
06-24 09:27:59.290 15688 15728 F libc    : Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0 in tid 15728 (Thread-6), pid 15688 (ValidationTests)
06-24 09:27:59.333 22488 22488 E crash_dump64: unknown process state: t
06-24 09:27:59.363 22488 22488 I crash_dump64: obtaining output fd from tombstoned, type: kDebuggerdTombstone
06-24 09:27:59.363  5588  5588 I /system/bin/tombstoned: received crash request for pid 15728
06-24 09:27:59.364 22488 22488 I crash_dump64: performing dump of process 15688 (target tid = 15728)
06-24 09:27:59.372 22488 22488 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
06-24 09:27:59.372 22488 22488 F DEBUG   : Build fingerprint: 'samsung/beyond1ltexx/beyond1:10/QP1A.190711.020/G973FXXU3BSKO:user/release-keys'
06-24 09:27:59.372 22488 22488 F DEBUG   : Revision: '26'
06-24 09:27:59.372 22488 22488 F DEBUG   : ABI: 'arm64'
06-24 09:27:59.372 22488 22488 F DEBUG   : Timestamp: 2020-06-24 09:27:59-0600
06-24 09:27:59.373 22488 22488 F DEBUG   : pid: 15688, tid: 15728, name: Thread-6  >>> com.example.VulkanLayerValidationTests <<<
06-24 09:27:59.373 22488 22488 F DEBUG   : uid: 11928
06-24 09:27:59.373 22488 22488 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0
06-24 09:27:59.373 22488 22488 F DEBUG   : Cause: null pointer dereference
06-24 09:27:59.373 22488 22488 F DEBUG   :     x0  0000000000000001  x1  0000000000000000  x2  0000000000000000  x3  0000000000000000
06-24 09:27:59.373 22488 22488 F DEBUG   :     x4  0000000000000000  x5  0000000000000000  x6  0000000000000000  x7  0000000000040000
06-24 09:27:59.373 22488 22488 F DEBUG   :     x8  0000000000000001  x9  0000000000000000  x10 0000006f9fc52a88  x11 0000000000000028
06-24 09:27:59.373 22488 22488 F DEBUG   :     x12 0201010102010100  x13 000000702b1df874  x14 000000702b1df8b8  x15 000000003b9d71b0
06-24 09:27:59.373 22488 22488 F DEBUG   :     x16 0000000000000001  x17 0000000001020101  x18 000000003b9cf897  x19 0000006fb2ca9fc8
06-24 09:27:59.373 22488 22488 F DEBUG   :     x20 000000702b1dfc80  x21 0000000000000000  x22 0000000000000000  x23 0000000000000000
06-24 09:27:59.373 22488 22488 F DEBUG   :     x24 0000006fb2ca9fc0  x25 000000702b1e1020  x26 0000006fb2ca9fc8  x27 000000702b1e1020
06-24 09:27:59.373 22488 22488 F DEBUG   :     x28 000000702b1df868  x29 000000702b1e00c0
06-24 09:27:59.373 22488 22488 F DEBUG   :     sp  000000702b1df850  lr  000000701c180870  pc  000000701c1809a8
06-24 09:27:59.376 22488 22488 F DEBUG   : 
06-24 09:27:59.376 22488 22488 F DEBUG   : backtrace:
06-24 09:27:59.376 22488 22488 F DEBUG   :       #00 pc 0000000000ab09a8  /vendor/lib64/egl/libGLES_mali.so (BuildId: 2074cca0c239c545a7bd2396c71003a9)
06-24 09:27:59.376 22488 22488 F DEBUG   :       #01 pc 0000000000ab05dc  /vendor/lib64/egl/libGLES_mali.so (vkCreateDevice+260) (BuildId: 2074cca0c239c545a7bd2396c71003a9)
06-24 09:27:59.376 22488 22488 F DEBUG   :       #02 pc 0000000000026f8c  /system/lib64/libvulkan.so (vulkan::driver::CreateDevice(VkPhysicalDevice_T*, VkDeviceCreateInfo const*, VkAllocationCallbacks const*, VkDevice_T**)+476) (BuildId: d569637329ba16330d390540dee0e601)
06-24 09:27:59.376 22488 22488 F DEBUG   :       #03 pc 00000000008dceb0  /data/app/com.example.VulkanLayerValidationTests-wa-4ADRfel00oEG07PSiRw==/lib/arm64/libVkLayer_khronos_validation.so (BuildId: 44934a49e4cdf81cc9e08014663b14d61afa4a41)
06-24 09:27:59.376 22488 22488 F DEBUG   :       #04 pc 00000000000164cc  /system/lib64/libvulkan.so (vulkan::api::CreateDevice(VkPhysicalDevice_T*, VkDeviceCreateInfo const*, VkAllocationCallbacks const*, VkDevice_T**)+1412) (BuildId: d569637329ba16330d390540dee0e601)
06-24 09:27:59.376 22488 22488 F DEBUG   :       #05 pc 00000000002fda38  /data/app/com.example.VulkanLayerValidationTests-wa-4ADRfel00oEG07PSiRw==/lib/arm64/libVulkanLayerValidationTests.so (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-24 09:27:59.376 22488 22488 F DEBUG   :       #06 pc 00000000003ab950  /data/app/com.example.VulkanLayerValidationTests-wa-4ADRfel00oEG07PSiRw==/lib/arm64/libVulkanLayerValidationTests.so (testing::Test::Run()+384) (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-24 09:27:59.376 22488 22488 F DEBUG   :       #07 pc 00000000003acc08  /data/app/com.example.VulkanLayerValidationTests-wa-4ADRfel00oEG07PSiRw==/lib/arm64/libVulkanLayerValidationTests.so (testing::TestInfo::Run()+396) (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-24 09:27:59.376 22488 22488 F DEBUG   :       #08 pc 00000000003ad1a8  /data/app/com.example.VulkanLayerValidationTests-wa-4ADRfel00oEG07PSiRw==/lib/arm64/libVulkanLayerValidationTests.so (testing::TestCase::Run()+248) (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-24 09:27:59.376 22488 22488 F DEBUG   :       #09 pc 00000000003b5f00  /data/app/com.example.VulkanLayerValidationTests-wa-4ADRfel00oEG07PSiRw==/lib/arm64/libVulkanLayerValidationTests.so (testing::internal::UnitTestImpl::RunAllTests()+1080) (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-24 09:27:59.376 22488 22488 F DEBUG   :       #10 pc 00000000003b5a94  /data/app/com.example.VulkanLayerValidationTests-wa-4ADRfel00oEG07PSiRw==/lib/arm64/libVulkanLayerValidationTests.so (testing::UnitTest::Run()+164) (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-24 09:27:59.376 22488 22488 F DEBUG   :       #11 pc 00000000001e09d4  /data/app/com.example.VulkanLayerValidationTests-wa-4ADRfel00oEG07PSiRw==/lib/arm64/libVulkanLayerValidationTests.so (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-24 09:27:59.376 22488 22488 F DEBUG   :       #12 pc 0000000000399aa8  /data/app/com.example.VulkanLayerValidationTests-wa-4ADRfel00oEG07PSiRw==/lib/arm64/libVulkanLayerValidationTests.so (BuildId: 67a6ccce404a07d99d3cd342b651e77cc2bc53c0)
06-24 09:27:59.376 22488 22488 F DEBUG   :       #13 pc 00000000000e3b14  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+36) (BuildId: 220051b49364b1c2da3adf10c30832cc)
06-24 09:27:59.376 22488 22488 F DEBUG   :       #14 pc 0000000000085330  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64) (BuildId: 220051b49364b1c2da3adf10c30832cc)
```

SzilardPapp-Arm commented:
@mark-lunarg, not quite sure why it fails on the S10.  I had problems with the multiple device creation/destruction, not sure what would happen if I just call vkDestroyDevice once in the end.
Do you have any ideas?

SzilardPapp-Arm commented:
I'm still not quite sure what's causing the crash on the S10.
There might be something with the multiple device creation/destruction, so I removed that first bit of the test when it was checking if the error message is not there by mistake and just left the test where it's checking if the warning message is raised when it should be raised.
I also removed the vk::DestroyDevice function, as it was not present in test "UseDeprecatedDeviceExtensions", so I thought it might not be necessary(It was also giving me access violation error when I left it in the end of the test).

mark-lunarg commented:
Running through CI again...

mark-lunarg commented on tests/vklayertests_best_practices.cpp:
```diff
@@ -1247,4 +1247,42 @@ TEST_F(VkArmBestPracticesLayerTest, PipelineDepthBiasZeroTest) {
+    if (supportedFeatures.robustBufferAccess) {
+        m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_PERFORMANCE_WARNING_BIT_EXT,
+                                             "UNASSIGNED-BestPractices-vkCreateDevice-RobustBufferAccess");
+        VkPhysicalDeviceFeatures deviceFeatures = {};
```
Fixed these local variables this time (this and supportedFeatures).

mark-lunarg commented:
@SzilardPapp-Arm, just a note -- every API call is controlled by boolean skip value returned from `LogError()` or `LogWarning()` and set by the debug callback.  The test harness will set `skip `to TRUE if the correct validation error/warning is detected in the test.  This, in turn, determines whether the down-chain call actually occurs, and allows us to make invalid `CreateDevice `calls (for instance) without the munged parameters making it all the way to the driver and crashing.

mark-lunarg commented:
Rebased, fixed up commit messages and var names, pushed to Master.
